### PR TITLE
bugfix: fail to delete pods due to CNI panic

### LINKED ISSUE #377: Calico CNI panics during pod deletion. Kubelet cannot delete the pod
cni-plugin panics when kubelet tries to delete a pod, so kubelet constantly keeps retrying resulting in pods stuck in deletion, or ContainerCreating state

## Expected Behavior

It should not panic.

## Current Behavior

It panics with the following output:

```bash

Sep 08 17:20:11 node-02 kubelet[2405]: E0908 17:20:11.423224    2405 cni.go:312] Error deleting network: netplugin failed but error parsing its diagnostic message "": unexpected end of JSON input
Sep 08 17:20:11 node-02 kubelet[2405]: E0908 17:20:11.424033    2405 remote_runtime.go:114] StopPodSandbox "0866c8b5892313d3e81696522380d7bae68ccbfc3281c6da267315c7564e685c" from runtime service failed: rpc error: code = 2 desc = NetworkP
Sep 08 17:20:11 node-02 kubelet[2405]: E0908 17:20:11.424063    2405 kuberuntime_manager.go:783] Failed to stop sandbox {"docker" "0866c8b5892313d3e81696522380d7bae68ccbfc3281c6da267315c7564e685c"}
Sep 08 17:20:11 node-02 kubelet[2405]: E0908 17:20:11.424099    2405 kuberuntime_manager.go:572] killPodWithSyncResult failed: failed to "KillPodSandbox" for "77bf230a-94b9-11e7-846b-080027ee3298" with KillPodSandboxError: "rpc error: cod
Sep 08 17:20:11 node-02 kubelet[2405]: E0908 17:20:11.424119    2405 pod_workers.go:182] Error syncing pod 77bf230a-94b9-11e7-846b-080027ee3298 ("busybox-383812631-w65hr_default(77bf230a-94b9-11e7-846b-080027ee3298)"), skipping: failed to
Sep 08 17:20:15 node-02 kubelet[2405]: I0908 17:20:15.078413    2405 kuberuntime_manager.go:901] updating runtime config through cri with podcidr 192.168.1.0/24
Sep 08 17:20:15 node-02 kubelet[2405]: I0908 17:20:15.078593    2405 docker_service.go:308] docker cri received runtime config &RuntimeConfig{NetworkConfig:&NetworkConfig{PodCidr:192.168.1.0/24,},}
Sep 08 17:20:15 node-02 kubelet[2405]: I0908 17:20:15.081028    2405 kubelet_network.go:326] Setting Pod CIDR:  -> 192.168.1.0/24
Sep 08 17:20:24 node-02 kubelet[2405]: I0908 17:20:24.342570    2405 kuberuntime_manager.go:457] Container {Name:busybox Image:busybox Command:[sleep 999999] Args:[] WorkingDir: Ports:[] EnvFrom:[] Env:[] Resources:{Limits:map[] Requests:
Sep 08 17:20:24 node-02 kubelet[2405]: W0908 17:20:24.345139    2405 cni.go:258] CNI failed to retrieve network namespace path: Cannot find network namespace for the terminated container "0866c8b5892313d3e81696522380d7bae68ccbfc3281c6da26
Sep 08 17:20:24 node-02 kubelet[2405]: 2017-09-08 17:20:24.408 [INFO][5933] calico.go 312: Extracted identifiers Node="node-02" Orchestrator="k8s" Workload="default.busybox-383812631-w65hr"
Sep 08 17:20:24 node-02 kubelet[2405]: 2017-09-08 17:20:24.408 [INFO][5933] utils.go 250: Configured environment: [LANG=en_US LANGUAGE=en_US: PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin KUBELET_KUBECONFIG_ARGS=--kube
Sep 08 17:20:24 node-02 kubelet[2405]: 2017-09-08 17:20:24.408 [INFO][5933] client.go 202: Loading config from environment
Sep 08 17:20:24 node-02 kubelet[2405]: 2017-09-08 17:20:24.418 [WARNING][5933] k8s.go 313: WorkloadEndpoint does not exist in the datastore, moving forward with the clean up Workload="default.busybox-383812631-w65hr" WorkloadEndpoint=api.
Sep 08 17:20:24 node-02 kubelet[2405]: Calico CNI releasing IP address
Sep 08 17:20:24 node-02 kubelet[2405]: 2017-09-08 17:20:24.418 [INFO][5933] utils.go 94: Using a dummy podCidr to release the IP Workload="default.busybox-383812631-w65hr" podCidr="0.0.0.0/0"
Sep 08 17:20:24 node-02 kubelet[2405]: panic: value method api.WorkloadEndpoint.String called using nil *WorkloadEndpoint pointer
Sep 08 17:20:24 node-02 kubelet[2405]: goroutine 1 [running, locked to thread]:
Sep 08 17:20:24 node-02 kubelet[2405]: github.com/projectcalico/cni-plugin/vendor/github.com/projectcalico/libcalico-go/lib/api.(*WorkloadEndpoint).String(0x0, 0x1458100, 0x0)
Sep 08 17:20:24 node-02 kubelet[2405]:         <autogenerated>:60 +0xf3
Sep 08 17:20:24 node-02 kubelet[2405]: github.com/projectcalico/cni-plugin/vendor/github.com/projectcalico/libcalico-go/lib/logutils.appendKVsAndNewLine(0xc4203fa000, 0xc420418c30)
Sep 08 17:20:24 node-02 kubelet[2405]:         /go/src/github.com/projectcalico/cni-plugin/vendor/github.com/projectcalico/libcalico-go/lib/logutils/logutils.go:114 +0x3bf
Sep 08 17:20:24 node-02 kubelet[2405]: github.com/projectcalico/cni-plugin/vendor/github.com/projectcalico/libcalico-go/lib/logutils.(*Formatter).Format(0x1f3b470, 0xc420418c30, 0xc42000a1a0, 0x15861e0, 0xc4203fa000, 0xc42026ed80, 0xc4201
Sep 08 17:20:24 node-02 kubelet[2405]:         /go/src/github.com/projectcalico/cni-plugin/vendor/github.com/projectcalico/libcalico-go/lib/logutils/logutils.go:71 +0x388
Sep 08 17:20:24 node-02 kubelet[2405]: github.com/projectcalico/cni-plugin/vendor/github.com/sirupsen/logrus.Entry.log(0xc420012230, 0xc42023efc0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc42037f204, ...)
Sep 08 17:20:24 node-02 kubelet[2405]:         /go/src/github.com/projectcalico/cni-plugin/vendor/github.com/sirupsen/logrus/entry.go:105 +0x1f1
Sep 08 17:20:24 node-02 kubelet[2405]: github.com/projectcalico/cni-plugin/vendor/github.com/sirupsen/logrus.(*Entry).Info(0xc420418be0, 0xc4207df5b0, 0x1, 0x1)
Sep 08 17:20:24 node-02 kubelet[2405]:         /go/src/github.com/projectcalico/cni-plugin/vendor/github.com/sirupsen/logrus/entry.go:140 +0xaa
Sep 08 17:20:24 node-02 kubelet[2405]: github.com/projectcalico/cni-plugin/k8s.CmdDelK8s(0xc42026a4f0, 0x0, 0x0, 0xc42000e05b, 0x4, 0xc4204783a0, 0x1f, 0x15c1f51, 0x3, 0xc420451870, ...)
Sep 08 17:20:24 node-02 kubelet[2405]:         /go/src/github.com/projectcalico/cni-plugin/k8s/k8s.go:367 +0x3c7
Sep 08 17:20:24 node-02 kubelet[2405]: main.cmdDel(0xc420299b90, 0xc4204510e0, 0x5)
Sep 08 17:20:24 node-02 kubelet[2405]:         /go/src/github.com/projectcalico/cni-plugin/calico.go:328 +0x9e8
Sep 08 17:20:24 node-02 kubelet[2405]: github.com/projectcalico/cni-plugin/vendor/github.com/containernetworking/cni/pkg/skel.(*dispatcher).checkVersionAndCall(0xc4201f44c0, 0xc420299b90, 0x1eaa300, 0xc4200185d0, 0x163d360, 0x0, 0xc4202cb
Sep 08 17:20:24 node-02 kubelet[2405]:         /go/src/github.com/projectcalico/cni-plugin/vendor/github.com/containernetworking/cni/pkg/skel/skel.go:162 +0x19f
Sep 08 17:20:24 node-02 kubelet[2405]: github.com/projectcalico/cni-plugin/vendor/github.com/containernetworking/cni/pkg/skel.(*dispatcher).pluginMain(0xc4201f44c0, 0x163d358, 0x163d360, 0x1eaa300, 0xc4200185d0, 0xc4207dff00)
Sep 08 17:20:24 node-02 kubelet[2405]:         /go/src/github.com/projectcalico/cni-plugin/vendor/github.com/containernetworking/cni/pkg/skel/skel.go:175 +0x2af
Sep 08 17:20:24 node-02 kubelet[2405]: github.com/projectcalico/cni-plugin/vendor/github.com/containernetworking/cni/pkg/skel.PluginMainWithError(0x163d358, 0x163d360, 0x1eaa300, 0xc4200185d0, 0xba)
Sep 08 17:20:24 node-02 kubelet[2405]:         /go/src/github.com/projectcalico/cni-plugin/vendor/github.com/containernetworking/cni/pkg/skel/skel.go:210 +0xed
Sep 08 17:20:24 node-02 kubelet[2405]: github.com/projectcalico/cni-plugin/vendor/github.com/containernetworking/cni/pkg/skel.PluginMain(0x163d358, 0x163d360, 0x1eaa300, 0xc4200185d0)
Sep 08 17:20:24 node-02 kubelet[2405]:         /go/src/github.com/projectcalico/cni-plugin/vendor/github.com/containernetworking/cni/pkg/skel/skel.go:222 +0x56
Sep 08 17:20:24 node-02 kubelet[2405]: main.main()
Sep 08 17:20:24 node-02 kubelet[2405]:         /go/src/github.com/projectcalico/cni-plugin/calico.go:384 +0x1e6
Sep 08 17:20:24 node-02 kubelet[2405]: E0908 17:20:24.424773    2405 cni.go:312] Error deleting network: netplugin failed but error parsing its diagnostic message "": unexpected end of JSON input
Sep 08 17:20:24 node-02 kubelet[2405]: E0908 17:20:24.426073    2405 remote_runtime.go:114] StopPodSandbox "0866c8b5892313d3e81696522380d7bae68ccbfc3281c6da267315c7564e685c" from runtime service failed: rpc error: code = 2 desc = NetworkP
Sep 08 17:20:24 node-02 kubelet[2405]: E0908 17:20:24.426115    2405 kuberuntime_manager.go:783] Failed to stop sandbox {"docker" "0866c8b5892313d3e81696522380d7bae68ccbfc3281c6da267315c7564e685c"}
``` 

## Steps to Reproduce (for bugs)

1. start a k8s cluster
2. add a new node without `podCIDR` label populated
3. schedule a pod
4. you'll see a message like this `rpc error: code = 2 desc = failed to start container "9182eb3c73bdab20764b6b1a5828f824cf20a921c598f418a548ee831a25a8e7": Error response from daemon: {"message":"cannot join network of a non running container: 0866c8b5892313d3e81696522380d7bae68ccbfc3281c6da267315c7564e685c"}` in `kubectl get pod`

## Your Environment

* CNI plugin version (e.g. `/opt/cni/bin/calico --version`): `v1.10.0`
* Orchestrator version (e.g. kubernetes, mesos, rkt): `k8s` with KDD or etcd backend
* CNI config: standard KDD/etcd manifest
* Operating System and version: ubuntu 16.04

### PR DESCRIPTION
**Pull request by gunjan5**
## Description

Fixes #377 

## Todos
- [ ] Tests
- [ ] Documentation
- [ ] Release note

## Release Note

```release-note
Bugfix: container deletion failure due to CNI panic
```

### CODE REVIEW

File: k8s/k8s.go [hunk 1]
```diff
@@ -364,7 +364,12 @@ func CmdDelK8s(c *calicoclient.Client, ep api.WorkloadEndpointMetadata, args *sk
 		return ipamErr
 	}
 
-	logger.WithField("endpoint", wep).Info("Teardown processing complete.")
```
Comment (caseydavenport) [inline k8s/k8s.go]: A nicer fix might be to add the `workloadID` (and probably container ID as well?) to the context logger here: https://github.com/projectcalico/cni-plugin/blob/master/calico.go#L82

And then just turn this into a single log "Teardown processing complete" - we log the wep above if it exists.

Comment (caseydavenport) [reply k8s/k8s.go]: wdyt?

Comment (gunjan5) [reply k8s/k8s.go]: hmm yeah, but I'm wondering if there's a bunch of pod deletion/addition logs will interleave and that will make it hard to know which WEP was torn down. Would that not be the case?

Comment (gunjan5) [reply k8s/k8s.go]: Actually, scratch that. `workload` is already part of the context logger for `cmdDel`, we should've just used that. I can also add it (and containerID) to the `cmdAdd` context logger as well

**Pull request merged**